### PR TITLE
Use winehq debs for win32 tests

### PR DESCRIPTION
**Pull request by ajtowns**
When running tests with wine 1.6, "make check" takes about 20 minutes;
with wine 3.0, "make check" takes about 7 minutes, allowing the travis
timeout to be lowered. Debs for wine 3.0 are available from winehq,
so this patch includes their repo key.

### CODE REVIEW

**Comment by laanwj** [comment]
I like using a recent wine. Unlike other kinds, winehq's wine doesn't age well.

This is their "wine-stable" so we don't (as we shouldn't) take upon us the job of testing experimental new wine, either.

Would even more prefer to bump entire test environment to a more recent distribution (ref #12511) but I don't think Travis makes that possible (or easy), so this is a good intermediate step.

utACK https://github.com/bitcoin/bitcoin/pull/12931/commits/a833fe6df5c86c7cf3bae319a000b70048031e31

Additional motivation (from IRC):
```
<aj> wumpus: ^^ is a first pass at stopping 686-w32 travis builds from being constantly horrible
```

**Comment by kallewoof** [comment]
`Timeout (15 minutes) reached.` -- I think non-Win jobs occasionally take up to 30 mins (for some reason), so this may be too low. That aside, utACK!

**Comment by practicalswift** [comment]
utACK a833fe6df5c86c7cf3bae319a000b70048031e31 module adjustment/removal of timeout change

File: .travis.yml [hunk 1]
```diff
@@ -78,7 +82,7 @@ script:
     - ./configure --cache-file=../config.cache $BITCOIN_CONFIG_ALL $BITCOIN_CONFIG || ( cat config.log && false)
     - make $MAKEJOBS $GOAL || ( echo "Build failure. Verbose build follows." && make $GOAL V=1 ; false )
     - export LD_LIBRARY_PATH=$TRAVIS_BUILD_DIR/depends/$HOST/lib
-    - if [ "$RUN_TESTS" = "true" ]; then travis_wait 50 make $MAKEJOBS check VERBOSE=1; fi
+    - if [ "$RUN_TESTS" = "true" ]; then travis_wait 15 make $MAKEJOBS check VERBOSE=1; fi
```
Comment (practicalswift) [inline .travis.yml]: What is the reasoning behind the change from 50 to 15 here?

Comment (kallewoof) [reply .travis.yml]: It was raised from 30 to 50. I think it should be dropped to 30 again.

Comment (practicalswift) [reply .travis.yml]: 30 makes sense

Comment (practicalswift) [reply .travis.yml]: 30 sounds reasonable.

**Comment by ajtowns** [comment]
@kallewoof Hmm, the timeout was for the x86-64 win32 build, which I left using wine1.6. Anyone have any examples of a non-win32 travis job where "make check" took longer than 5 or 10 minutes?

**Comment by laanwj** [comment]
> @kallewoof Hmm, the timeout was for the x86-64 win32 build, which I left using wine1.6. 

I think that's confusing - let's switch all the wine builds around.

**Comment by MarcoFalke** [comment]
Needs rebase if still relevant

**Comment by MarcoFalke** [comment]
Seems to give a speedup of 2

**Comment by jnewbery** [comment]
Concept ACK.

The `apt-get install` step takes ~ 6 minutes (https://travis-ci.org/bitcoin/bitcoin/jobs/364620673#L494). Is it possible to cache the dependency in Travis (perhaps using https://docs.travis-ci.com/user/caching#Arbitrary-directories)?

**Comment by MarcoFalke** [comment]
Ouch. That dwarfs any speedups gained later on. I'd highly suspect the bottleneck is installing the ~200 packages and not fetching them, so caching woulnd't help here.

**Comment by jamesob** [comment]
> I'd highly suspect the bottleneck is installing the ~200 packages and not fetching them, so caching woulnd't help here.

That seems to somewhat contradict [what Travis' documentation says on what to cache](https://docs.travis-ci.com/user/caching#Things-not-to-cache):

> Large files that are quick to install but slow to download do not benefit from caching, as they take as long to download from the cache as from the original source

Why not try it?

**Comment by MarcoFalke** [comment]
> Edit: ah, they specifically call out "Debian packages" below as something not to cache. Oops :).

Yeah, I'd rather not install those. I'd surprised if all of them are actually needed.

**Comment by kallewoof** [comment]
Could the `--install-recommends` part be fine tuned (replace with explicit package names) to exclude unnecessary stuff?

**Comment by laanwj** [comment]
> Could the --install-recommends part be fine tuned (replace with explicit package names) to exclude unnecessary stuff?

Good idea. `--install-recommends` is almost always a bad idea. As `test_bitcoin` is a basic console program I wouldn't be surprised if we only need the wine base package.

**Comment by laanwj** [comment]
Needs update to commit message and PR title now that all windows tests are performed using the new wine.

**Comment by jnewbery** [comment]
Much better! apt-get is now taking 6 seconds on the i686-w64-mingw32 job and 30 seconds on the x86_64-w64-mingw32 job.

**Comment by ajtowns** [comment]
PR title updated, commit message had already been updated. Even with `--no-install-recommends`, it's grabbing a lot of packages (100MB versus 130MB by the looks), and it also seems like travis's download speeds can be wildly inconsistent:

* `Fetched 109 MB in 2s (51.9 MB/s)` -- https://travis-ci.org/bitcoin/bitcoin/jobs/364960356
* `Fetched 133 MB in 5min 9s (429 kB/s)` -- https://travis-ci.org/bitcoin/bitcoin/jobs/364620673
* `Fetched 133 MB in 2s (47.3 MB/s)` -- https://travis-ci.org/bitcoin/bitcoin/jobs/364574088

**Comment by MarcoFalke** [comment]
utACK e5997988149b7ecf41e73815c6b161b3c1dff1b4 (Would still be an improvement if it took travis 6 minutes to install winehq)

previous `make check` times would be between 915 and 1350 and new times between 405 and 530, so as long as we spend less than 6 minutes in `apt-get` it is a net win.

**Comment by MarcoFalke** [comment]
Just reset the travis run and it seems the slow fetch was a one-off. Changes look good to me...

**Comment by jnewbery** [comment]
@MarcoFalke 
> I'd highly suspect the bottleneck is installing the ~200 packages and not fetching them

@ajtowns 
>  it also seems like travis's download speeds can be wildly inconsistent:

It appears that fetching the packages *is* the bottleneck here, which suggests that caching locally would help. That could be investigated in a follow-up PR - no need to solve it here if this is already a net win.

**Comment by MarcoFalke** [comment]
Yeah, right. Though, downloading the packages from the mirror shouldn't take 5 minutes either. Note that the cache is not stored internally but on S3 [1], so it might be further away from gce (where our vms run) than the gce mirror of the ubuntu packages.

[1] https://docs.travis-ci.com/user/caching#How-does-the-caching-work%3F

**Comment by theuni** [comment]
Has anyone tracked down exactly _what_ is sped up by newer Wine? I'm wondering if there's some one-off test that accounts for the bulk of the difference.

**Comment by theuni** [comment]
See #12970, which may negate some of the gains here.

**Comment by ajtowns** [comment]
Rebased on top of the now-merged #12970. I think at this point it's not worth the added risk/complexity of pulling in out-of-tree packages, but it does seem interesting to see if there's any further improvement. :)

**Comment by MarcoFalke** [comment]
Can be closed?

**Comment by ajtowns** [comment]
I think so. I suspect the "running slower" is just variance from travis, but either way any gains do seem small enough not to be worthwhile. wine-stable 3.0 is in bionic as well.

**Pull request closed without merging**